### ISSUE TITLE
Setting SCOUT_DATA_FILE config breaks ScoutApm reporting

### ISSUE THREAD

**Issue by kainosnoema** [scoutapp/scout_apm_ruby#59]
When the `data_file` config setting is set, ScoutApm agent cannot report because there's a [bug in `Layaway` directory initialization](https://github.com/scoutapp/scout_apm_ruby/blob/48b56833d02d22cdf55ff18793f5034470e22051/lib/scout_apm/layaway.rb#L34):

```
/srv/homebase/current/vendor/bundle/ruby/2.1.0/gems/scout_apm-2.0.0/lib/scout_apm/layaway.rb:34:in `directory?': wrong number of arguments (0 for 1) (ArgumentError)
	from /srv/homebase/current/vendor/bundle/ruby/2.1.0/gems/scout_apm-2.0.0/lib/scout_apm/layaway.rb:34:in `directory'
	from /srv/homebase/current/vendor/bundle/ruby/2.1.0/gems/scout_apm-2.0.0/lib/scout_apm/layaway.rb:125:in `glob_pattern'
	from /srv/homebase/current/vendor/bundle/ruby/2.1.0/gems/scout_apm-2.0.0/lib/scout_apm/layaway.rb:112:in `file_for'
	from /srv/homebase/current/vendor/bundle/ruby/2.1.0/gems/scout_apm-2.0.0/lib/scout_apm/layaway.rb:48:in `write_reporting_period'
	from /srv/homebase/current/vendor/bundle/ruby/2.1.0/gems/scout_apm-2.0.0/lib/scout_apm/store.rb:74:in `block (2 levels) in write_to_layaway'
	from /srv/homebase/current/vendor/bundle/ruby/2.1.0/gems/scout_apm-2.0.0/lib/scout_apm/store.rb:72:in `each'
	from /srv/homebase/current/vendor/bundle/ruby/2.1.0/gems/scout_apm-2.0.0/lib/scout_apm/store.rb:72:in `block in write_to_layaway'
	from /srv/homebase/current/vendor/bundle/ruby/2.1.0/gems/scout_apm-2.0.0/lib/scout_apm/store.rb:70:in `synchronize'
	from /srv/homebase/current/vendor/bundle/ruby/2.1.0/gems/scout_apm-2.0.0/lib/scout_apm/store.rb:70:in `write_to_layaway'
	from /srv/homebase/current/vendor/bundle/ruby/2.1.0/gems/scout_apm-2.0.0/lib/scout_apm/agent.rb:205:in `shutdown'
	from /srv/homebase/current/vendor/bundle/ruby/2.1.0/gems/scout_apm-2.0.0/lib/scout_apm/agent.rb:189:in `block in install_exit_handler'
```

I tried fixing this to submit a PR, but couldn't get past a series of issues when bootstrapping the project and running tests. Couple issues I ran into there:

1) `LoadError: cannot load such file -- scout_apm/utils/fake_stack_prof`. There's no file matching this in the repo right now.
2) Many undefined variables and methods, including (just two out of many):

```
NameError: undefined local variable or method `agent' for #<SlowRequestPolicyTest:0x007f8a4d1fd068>
    /Users/evan/Code/cotap/scout_apm_ruby/test/unit/slow_request_policy_test.rb:35:in `test_score'
```

and

```
NoMethodError: undefined method `add_reporting_period' for #<ScoutApm::Layaway:0x007f8a4e073890 @directory=nil>
    /Users/evan/Code/cotap/scout_apm_ruby/test/unit/layaway_test.rb:22:in `test_merge_reporting_period'
```

What am I missing? Tips on getting the tests in master passing?

**Comment by kainosnoema** [comment]
By the way, this config value isn't documented—I just found it while looking for a solution to another problem we ran into: because the agent writes its layaway data files to the Rails root by default, these new files trigger a re-deploy loop in our Chef-based artifact deploys. We can/should probably symlink `tmp/` from a shared location into the root so any changes there don't trigger a re-deploy, but changing the Scout data file location to `/tmp` seemed like a quick short-term fix.

**Comment by cschneid** [comment]
@kainosnoema:

1. I released scout_apm version 2.1.1, with a fix for the immediate issue
2. I pruned and cleaned up the unit tests. They should all be passing now. Still looking into setting up a CI system. 

Thank you a ton for the report.

**Comment by kainosnoema** [comment]
@cschneid thanks for this!

By the way, in the course of resolving this we actually discovered a fairly serious side-affect when the agent can't write to the data file. Because there's no exception handling when the write fails, [the data during the reporting period is never removed](https://github.com/scoutapp/scout_apm_ruby/blob/7bcb83883c88f310530ac37d48f4a42a0f2fbd26/lib/scout_apm/store.rb#L74-L75). It just keeps accumulating in memory indefinitely causing a memory leak and eventually the process OOMs. Additionally, since the writing is done in a background thread and the thread isn't set to abort on exceptions, there's no external evidence that this is happening outside of memory usage.

This actually caused us some lost requests yesterday and today until we figured out that it was Scout still running but not able to write out the layaway files—in retrospect we should have disabled Scout entirely, but it's a vulnerability to be aware of and potentially handle more gracefully.
